test/functional/page_objects/discover_page.js

```diff
@@ -53,6 +53,10 @@ export function DiscoverPageProvider({ getService, getPageObjects }) {
       await getRemote().findDisplayedById('SaveSearch').pressKeys(searchName);
       await testSubjects.click('discoverSaveSearchButton');
       await PageObjects.header.waitUntilLoadingHasFinished();
+
+      // Make sure toast exists and then close it.
+      await testSubjects.existOrFail('saveSearchSuccess');
+      await find.clickByCssSelector('[data-test-subj="saveSearchSuccess"] [data-test-subj="toastCloseButton"]');
```

Comment:
All of the `testSubjects.existOrFails` are unnecessary because the find will fail if it doesn't find toast.  Or do you prefer keeping them in for readability?  Only concern is that they do disappear after awhile.  I don't *think* it will cause issues but it would stink if it passed the first call, then failed on the second because in the interim, the toast disappeared.  Unless we could do what we did with the old style toasts and modify them so they never disappear.  Or maybe we have that?  I'm not sure if the new toasts use the same advanced settings for how long to appear.

Reply:
If I were reading the code and I saw this on its own:

```js
await testSubject.click('saveSearchSuccess toastCloseButton');
```

It would not be clear to me why we're closing a toast that we haven't even looked at yet. The code implies the toast exists, but it doesn't verify this. I might even remove this code thinking that cleanup doesn't matter in this case, completely overlooking the role it plays as an assertion.

For me, the intent behind the code is clearer with two explicit actions: 1) the verification of success and 2) the removal of the toast to clean up the UI.

> Only concern is that they do disappear after awhile

They'll disappear after a few seconds, but as soon as `existsOrFail` finds the toast, doesn't it resolve the promise? So there won't be any delay between these two actions, right?

Reply:
> They'll disappear after a few seconds, but as soon as existsOrFail finds the toast, doesn't it resolve the promise? 

yes

> So there won't be any delay between these two actions, right?

Well, there is still a delay, just a very minor one, like, whatever time it takes the code to execute the next line.

> The code implies the toast exists, but it doesn't verify this.

We don't add manual assertions that any given button exists, for instance, before we click it.  Actually, we kind of do, but it's part of the click function, so we don't have to add a manual check before every click call.  What this is doing is adding that manual call that already exists inside the click function.

>  I might even remove this code thinking that cleanup doesn't matter in this case, completely overlooking the role it plays as an assertion.

sgtm.  We seem to be doing well with flakiness so maybe this doesn't do anything extra.

I guess the way I see it is the only thing this code does is add a very slight potential for extra flakiness by having a check for "existsOrFail" and then the subsequent click.  What if, for instance, the "existsOrFail" function ends up searching through a bunch of DOM elements before it gets to the toast, so that function call takes a few seconds to return true, then before the next line can execute, the toast disappears. Unlikely (especially because the existsOfFail check takes max 1 s), but maybe possible?

Reply:
> We don't add manual assertions that any given button exists, for instance, before we click it. Actually, we kind of do, but it's part of the click function, so we don't have to add a manual check before every click call.

I think I understand how you're approaching this. I think you're looking at this from the angle of "How do we remove points of failure", which I totally get. I'm looking at it from the point of view of "How do we make this code clear to the reader", which is an orthogonal concern to yours.

> What if, for instance, the "existsOrFail" function ends up searching through a bunch of DOM elements before it gets to the toast, so that function call takes a few seconds to return true, then before the next line can execute, the toast disappears.

I think what you're saying makes sense. I think this is possible and it's a valid concern. Though I'm not sure what's the best way to satisfy both of our concerns.

Reply:
> The code implies the toast exists, but it doesn't verify this. I might even remove this code thinking that cleanup doesn't matter in this case, completely overlooking the role it plays as an assertion.

Sorry, I read this wrong earlier.  I thought you were saying you were actually going to remove the code you added because it doesn't matter.

I'm confused by your intentions with this PR.  Initially I thought your goal was to avoid flakiness. 
 Then after re-reading your comment above, I thought your goal was to extend the tests.  Now I'm re-reading the primary git issue comment, and it's indeed to fix a flaky test?  

If you want to extend the tests to make them more robust (fail more often than they otherwise might now), I can see why you would add both lines (primarily want to make sure the toast is there, secondarily I want to dismiss it).

Or are you trying to reduce flakiness by first making sure the toast exists before dismissing it? In which case, the first line is pointless.

I'm just confused by what actually fixes the flakiness of the other test.  You wrote

> Fixes #20810 by asserting that the search has saved by checking for the presence of a success toast. 

But that doesn't make sense to me, why would adding a more thorough check cause it to pass?  If it does pass, it seems like it's accidental (like the extra check is similar to adding a little sleep and hence, you don't hit whatever you were hitting before).

I suppose it doesn't too much matter.  You can check this code in and maybe it'll accidentally help with stability, but it seems like there is an underlying issue that might show up later.